### DISCUSSION TITLE
Option to exclude currency symbol from currency method

### DISCUSSION THREAD

**Discussion by sts-ryan-holton** [laravel/framework#53017] [Ideas]
I've recently started using the [currency](https://laravel.com/docs/11.x/helpers#method-number-currency) method of the Number helper. In my use case I need to commit the currency symbol but retain the locale and other currency features. I tried passing an empty value to this but id doesn't return without a symbol.

Maybe we could introduce a new feature to exclude the symbol, or a new method that wraps the function just without the symbol?

```php
Number::currency(100, ''); // passing empty currency doesn't return without symbol
```

**Comment by ghost** [comment]
You can achieve your goal of formatting currency without displaying the currency symbol while retaining locale and other features by creating a custom wrapper around the `Number::currency` method. While the built-in method does not allow for suppressing the symbol directly, you can easily extend its functionality.

Here's how you could implement a custom helper function to format the number as currency without the symbol:

### Custom Helper Function

1. **Create a Custom Helper**

You can create a new helper function in a file such as `app/Helpers/NumberHelper.php` (or wherever you store your helpers):

```php
namespace App\Helpers;

use Illuminate\Support\Facades\Number;

class NumberHelper
{
    public static function currencyWithoutSymbol($value, $currency = null, $locale = null)
    {
        // Format the number as currency
        $formatted = Number::currency($value, $currency, $locale);

        // Remove the currency symbol
        return preg_replace('/[^\d.,]/', '', $formatted);
    }
}
```

### Usage

2. **Using the Custom Helper**

Now, you can call this helper wherever you need to format the number without a currency symbol:

```php
use App\Helpers\NumberHelper;

$amount = 100;
$formattedAmount = NumberHelper::currencyWithoutSymbol($amount, 'USD', 'en_US'); // returns "100.00"
```

### Registering the Helper

If you haven't already, ensure your helper file is autoloaded. You can do this by including it in `composer.json` under the `autoload` section:

```json
"autoload": {
    "files": [
        "app/Helpers/NumberHelper.php"
    ]
}
```

Then run:

```bash
composer dump-autoload
```

### Additional Considerations

- **Locale and Formatting:** This method retains locale formatting. If you pass different locales, it will adjust the decimal and thousands separators accordingly.
- **Performance:** Ensure that the regex used for stripping the currency symbol is performant for your use case.
- **Testing:** Test this function with various currency formats to ensure it meets your requirements.

### Alternative Approach

If you'd prefer to add this functionality directly into your existing application without a helper, consider extending the `Number` facade:

```php
namespace App\Facades;

use Illuminate\Support\Facades\Number as BaseNumber;

class Number extends BaseNumber
{
    public static function currencyWithoutSymbol($value, $currency = null, $locale = null)
    {
        $formatted = parent::currency($value, $currency, $locale);
        return preg_replace('/[^\d.,]/', '', $formatted);
    }
}
```

In this case, you’d register your custom facade in `config/app.php` and use it similarly as shown above.

**Reply by sts-ryan-holton** [reply]
@tal7aouy I could create a PR for this into the framework?

**Reply by shaedrich** [reply]
> ```php
> return preg_replace('/[^\d.,]/', '', $formatted);
> ```

This would also strip away other locale-specific formatting such as grouping, thousands separator, etc.

<small>(_BTW: Your answer is <del>written</del><ins>generated</ins> by AI, isn't it?_)</small>

If you wanted to keep these, this would result in something like this (and this only respects a few possibilities, yet there are more):

```php
namespace App\Helpers;

use Illuminate\Support\Facades\Number;

class NumberHelper
{
  private static function withLocale(array|string $locale, callable $callback)
  {
    $orginalMonetaryLocale = setlocale(LC_MONETARY, 0);
    $orginalNumericLocale = setlocale(LC_NUMERIC, 0);

    if (is_string($locale)) {
      $temporaryMonetaryLocale = $temporaryNumericLocale = $locale;
    } else {
      [ $temporaryMonetaryLocale, $temporaryNumericLocale ] = $locale;
    }

    if ($orginalMonetaryLocale !== $locale) {
      setlocale(LC_MONETARY, $locale);
    }

    if ($orginalNumericLocale !== $locale) {
      setlocale(LC_NUMERIC, $locale);
    }

    $result = call_user_func($callback, localeconv());

    if ($monetaryLocale !== $locale) {
      setlocale(LC_MONETARY, $orginalMonetaryLocale);
    }

    if ($orginalNumericLocale !== $locale) {
      setlocale(LC_NUMERIC, $orginalNumericLocale);
    }

    return $result;
  }

  public static function currencyWithoutSymbol($value, $currency = null, $locale = null)
  {
    return self::withLocale($locale, function(array $lConf) use ($value, $currency, $locale) {
      $reAllowList = [
        '\d',
        $lConf['mon_decimal_point'],
        $lConf['mon_thousands_sep'],
        $lConf['positive_sign'],
        $lConf['negative_sign'],
      ];

      // Format the number as currency
      $formatted = Number::currency($value, $currency, $locale);

      // Remove the currency symbol
      return preg_replace('/[^'.implode('', $reAllowList).']/', '', $formatted);
    });
  }
}
```

**Comment by shaedrich** [comment]
Without the currency symbol, it works pretty much like `Number::format()`, doesn't it? Doesn't that work for your use case?

Maybe, it's interesting to not, that the class is merely a wrapper for the native [`NumberFormatter`](https://www.php.net/manual/en/numberformatter.formatcurrency.php) and therefore aims to work quite similar. `NumberFormatter::formatCurrency()` in return is based on [ICU NumberFormat](https://unicode-org.github.io/icu/userguide/format_parse/numbers/).

You can probably create your own rule-based formatter, but that might require you to reproduce most of the existing currency formatting functionality.
